File: patterns/04-concurrency-patterns/06-structured-concurrency-cancellation/python/structured_concurrency_cancellation_test.py
from threading import Barrier, Event, Lock, Thread

from contract_support import run_contract
from structured_concurrency_cancellation import evaluate


def test_shared_contract():
    run_contract(__file__, evaluate)


def test_failure_cancels_sibling_and_all_threads_join():
    start = Barrier(3)
    cancelled = Event()
    lock = Lock()
    events = []

    def failing_child():
        start.wait()
        with lock:
            events.append("failed")
        cancelled.set()

    def sibling():
        start.wait()
        cancelled.wait()
        with lock:
            events.append("cancelled")

    threads = [Thread(target=failing_child), Thread(target=sibling)]
    for thread in threads:
        thread.start()
    start.wait()
    for thread in threads:
        thread.join()

    assert events == ["failed", "cancelled"]
    assert all(not thread.is_alive() for thread in threads)

File: patterns/04-concurrency-patterns/06-structured-concurrency-cancellation/js/structured_concurrency_cancellation.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import {runContract} from "../../../support/js/contract-support.mjs";
import {evaluate} from "./structured_concurrency_cancellation.mjs";

test("concurrency.structured-concurrency-cancellation shared contract", async () => {
    await runContract(import.meta.url, evaluate);
});

test("failure cancels sibling and all promises join", async () => {
    let releaseStart;
    let releaseCancellation;
    const start = new Promise((resolve) => { releaseStart = resolve; });
    const cancelled = new Promise((resolve) => { releaseCancellation = resolve; });
    const events = [];

    const failing = (async () => {
        await start;
        events.push("failed");
        releaseCancellation();
    })();
    const sibling = (async () => {
        await start;
        await cancelled;
        events.push("cancelled");
    })();

    releaseStart();
    await Promise.all([failing, sibling]);
    assert.deepEqual(events, ["failed", "cancelled"]);
});
